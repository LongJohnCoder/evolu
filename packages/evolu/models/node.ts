import nanoid from 'nanoid';
import { IO } from 'fp-ts/lib/IO';
import { Newtype, iso } from 'newtype-ts';

/**
 * NodeID is string created with nanoid.
 */
export interface NodeID
  extends Newtype<{ readonly NodeID: unique symbol }, string> {}

const isoNodeID = iso<NodeID>();

export const mapNodeIDToString = (nodeID: NodeID): string =>
  isoNodeID.unwrap(nodeID);

export const id: IO<NodeID> = () => isoNodeID.wrap(nanoid());

export interface Node {
  readonly id: NodeID;
}
