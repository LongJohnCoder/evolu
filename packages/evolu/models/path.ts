import { foldRight, getEq } from 'fp-ts/lib/Array';
import { eqNumber } from 'fp-ts/lib/Eq';
import { Endomorphism } from 'fp-ts/lib/function';
import { fromCompare, Ord } from 'fp-ts/lib/Ord';
import { pipe } from 'fp-ts/lib/pipeable';

export type Path = number[];

export const eqPath = getEq(eqNumber);

export const byDirection: Ord<Path> = fromCompare((x, y) =>
  eqPath.equals(x, y)
    ? 0
    : !x.some((value, index) => value > y[index])
    ? 1
    : -1,
);

export function movePath(offset: number): Endomorphism<Path> {
  return path =>
    pipe(
      path,
      foldRight(() => path, (init, last) => [...init, last + offset]),
    );
}
