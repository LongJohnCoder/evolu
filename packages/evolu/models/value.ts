import { Endomorphism } from 'fp-ts/lib/function';
import {
  alt,
  exists,
  fold,
  fromNullable,
  getOrElse,
  map,
  none,
  Option,
  some,
} from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { Lens } from 'monocle-ts';
import { createElement } from 'react';
import {
  Element,
  jsx,
  normalizeElement,
  ReactElement,
  setTextElement,
} from './element';
import { eqSelection, moveSelection, Selection } from './selection';

export interface Value {
  readonly element: Element;
  readonly hasFocus: boolean;
  readonly selection: Option<Selection>;
}

export const createValue = <T extends Element>({
  element,
  selection = none,
  hasFocus = false,
}: {
  element: T;
  selection?: Option<Selection>;
  hasFocus?: boolean;
}): Value => ({
  element,
  selection,
  hasFocus,
});

export const createValueWithText = (text = '') =>
  createValue<ReactElement>({
    element: jsx(createElement('div', { className: 'root' }, text)),
    hasFocus: false,
  });

/**
 * Focus on the element of Value.
 */
export const elementLens = Lens.fromProp<Value>()('element');

export const normalize: Endomorphism<Value> = value => {
  // https://github.com/gcanti/fp-ts/issues/976
  const element = normalizeElement(value.element);
  if (element === value.element) return value;
  return { ...value, element };
};

export const select = (selection: Selection): Endomorphism<Value> => value =>
  pipe(
    value.selection,
    exists(s => eqSelection.equals(s, selection)),
  )
    ? value
    : { ...value, selection: some(selection) };

export const setText = (
  text: string,
  selection?: Selection,
): Endomorphism<Value> => value =>
  pipe(
    fromNullable(selection),
    alt(() => value.selection),
    // Simple pipe nesting is ok. We can always refactor it out.
    map(selection =>
      pipe(
        value,
        elementLens.modify(setTextElement(text, selection)),
      ),
    ),
    getOrElse(() => value),
  );

// TODO: It should traverse across nodes.
export const move = (offset: number): Endomorphism<Value> => value =>
  pipe(
    value.selection,
    map(moveSelection(offset)),
    fold(() => value, selection => select(selection)(value)),
  );
