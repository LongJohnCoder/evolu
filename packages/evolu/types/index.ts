import { IO } from 'fp-ts/lib/IO';
import { Option } from 'fp-ts/lib/Option';
import { Task } from 'fp-ts/lib/Task';
import { Newtype } from 'newtype-ts';
import { ReactDOM, ReactNode, Reducer } from 'react';
import { $Values } from 'utility-types';
import {
  DOMNode,
  DOMRange,
  DOMSelection,
  DOMNodeOffset,
  ExistingDOMSelection,
} from './dom';

/**
 * Editor text is a string. Like in React.
 */
export type Text = string;

/**
 * Editor unique string ID generated by nanoid. ID is required for
 * meta-data, React keys, and CRDT.
 */
export interface ElementID
  extends Newtype<{ readonly ElementID: unique symbol }, string> {}

/**
 * Editor element. The base for all other editor elements.
 */
export interface Element {
  readonly id: ElementID;
  readonly children: Node[];
}

export type Node = Element | Text;

// TODO: Consider newtype.
export type PathIndex = number;

/**
 * Path to a place in Element. It can point to Element, Text, Text char, or nothing.
 */
export type Path = PathIndex[];

/**
 * Non empty Path.
 */
export interface NonEmptyPath extends Path {
  0: PathIndex;
}

/**
 * Non empty Path with offset.
 */
export interface NonEmptyPathWithOffset extends Path {
  0: PathIndex;
  1: PathIndex;
}

/**
 * Editor selection. It's like DOM Selection, but with Path for the anchor and the focus.
 * https://developer.mozilla.org/en-US/docs/Web/API/Selection
 */
export interface Selection {
  readonly anchor: NonEmptyPath;
  readonly focus: NonEmptyPath;
}

/**
 * Editor value.
 */
export interface Value {
  readonly element: Element;
  readonly hasFocus: boolean;
  readonly selection: Option<Selection>;
}

export type SetTextArg = {
  text: Text;
  path: NonEmptyPath;
  selection: Selection;
};

export type EditorAction =
  | { type: 'focus' }
  | { type: 'blur' }
  | { type: 'selectionChange'; selection: Selection }
  | { type: 'setText'; arg: SetTextArg }
  | { type: 'deleteContent'; selection: Selection };

export type EditorReducer = Reducer<Value, EditorAction>;

/**
 * Editor range. It's like DOM Range, but with editor path for the start and the end.
 * Range should be an implementation detail when an operation needs the direction.
 * https://developer.mozilla.org/en-US/docs/Web/API/Range
 */
export interface Range {
  readonly start: NonEmptyPath;
  readonly end: NonEmptyPath;
}

export type GetDOMNodeByPath = (path: Path) => IO<Option<DOMNode>>;
export type GetPathByDOMNode = (node: DOMNode) => IO<Option<Path>>;

export type SetDOMNodePathRef = (node: DOMNode | null) => void;

export type SetDOMNodePath = (
  operation: 'add' | 'remove',
  node: DOMNode,
  path: Path,
) => void;

export type RenderElement = (
  element: Element,
  children: ReactNode,
  ref: SetDOMNodePathRef,
) => ReactNode;

interface ReactElementFactory<T, P> extends Element {
  readonly tag: T;
  readonly props?: P;
  readonly children: (ReactElement | Text)[];
}

/**
 * Editor React-like element. It has tag and props.
 */
export type ReactElement = $Values<
  {
    [T in keyof ReactDOM]: ReactElementFactory<
      T,
      ReturnType<ReactDOM[T]>['props']
    >;
  }
>;

export type EditorElementAttrs = Pick<
  React.HTMLAttributes<HTMLDivElement>,
  | 'accessKey'
  | 'autoCorrect'
  | 'className'
  | 'id'
  | 'role'
  | 'spellCheck'
  | 'style'
  | 'tabIndex'
>;

export type EditorProps = EditorElementAttrs & {
  readonly value: Value;
  readonly onChange: (value: Value) => void;
  readonly renderElement?: RenderElement;
};

/**
 * Editor side effects. There are two abstractions meant for side effects:
 * IO (synchronous) and Task (asynchronous). Both of them have the -Either
 * "version" for computations that may fail.
 */
export interface EditorIO {
  readonly afterTyping: Task<void>;
  readonly createDOMRange: IO<Option<DOMRange>>;
  readonly createInfo: (selection: Selection) => Info; // TODO: IO
  readonly dispatch: (action: EditorAction) => IO<void>;
  readonly DOMRangeToSelection: (range: DOMRange) => IO<Option<Selection>>;
  readonly ensureDOMSelectionIsActual: IO<void>;
  readonly focus: IO<void>;
  readonly getDocument: IO<Option<Document>>;
  readonly getDOMNodeByPath: GetDOMNodeByPath;
  readonly getDOMSelection: IO<Option<DOMSelection>>;
  readonly getElement: IO<Option<HTMLDivElement>>;
  readonly getExistingDOMSelection: IO<Option<ExistingDOMSelection>>;
  readonly getPathByDOMNode: GetPathByDOMNode;
  readonly getSelectionFromDOM: IO<Option<Selection>>;
  readonly getValue: IO<Value>;
  readonly isTyping: IO<boolean>;
  readonly pathToNodeOffset: (path: NonEmptyPath) => IO<Option<DOMNodeOffset>>;
  readonly setDOMSelection: (selection: Selection) => IO<void>;
}

// TODO: Fragment, probably Child[].

export interface NodeInfo {
  readonly node: Node;
  readonly path: NonEmptyPath;
  // readonly text: string;
  // readonly parents: NonEmptyArray<Element>;
  // readonly parentBlocks: NonEmptyArray<Element>;
  // readonly previousSibling: Option<Child>;
  // readonly nextSibling: Option<Child>;
  // readonly textOffset: Option<number>;
  // readonly allChildrenCount: number;
}

/**
 * Info is materialized selection. It provides useful computations for
 * toolbars and operations.
 */
export interface Info {
  // selection: Selection;
  // range: Range;
  nodes: NodeInfo[];
  // text: Node;
  // range position
}
