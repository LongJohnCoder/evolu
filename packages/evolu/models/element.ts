import { foldRight, last, unsafeUpdateAt, init } from 'fp-ts/lib/Array';
import { Endomorphism, Predicate, Refinement } from 'fp-ts/lib/function';
import {
  chain,
  fold,
  fromPredicate,
  none,
  Option,
  some,
  toNullable,
} from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { Lens, Optional, Prism } from 'monocle-ts/lib';
import { indexArray } from 'monocle-ts/lib/Index/Array';
import { Children, ReactDOM, ReactNode } from 'react';
import { $Values } from 'utility-types';
import { EditorNode, id, SetNodeEditorPathRef } from './node';
import { EditorPath } from './path';
import {
  EditorSelection,
  editorSelectionAsRange,
  editorSelectionIsCollapsed,
} from './selection';
import { EditorText, editorTextIsBR, isEditorText } from './text';

/**
 * EditorElement is the base model for all other editor elements.
 */
export interface EditorElement extends EditorNode {
  readonly children: (EditorElementChild)[];
}

export type EditorElementChild = EditorElement | EditorText;

export type RenderEditorElement = (
  element: EditorElement,
  children: ReactNode,
  ref: SetNodeEditorPathRef,
) => ReactNode;

export const isEditorElement: Refinement<EditorNode, EditorElement> = (
  value,
): value is EditorElement => Array.isArray((value as EditorElement).children);

export const editorElementChildIsEditorElement: Refinement<
  EditorElementChild,
  EditorElement
> = (value): value is EditorElement => isEditorElement(value);

export const editorElementChildIsEditorText: Refinement<
  EditorElementChild,
  EditorText
> = (value): value is EditorText => isEditorText(value);

export type EditorTextWithOffset = {
  readonly editorText: EditorText;
  readonly offset: number;
};

export interface MaterializedEditorPath {
  to: EditorElement | EditorText | EditorTextWithOffset;
  parents: EditorElement[];
}

export const isEditorTextWithOffset: Refinement<
  MaterializedEditorPath['to'],
  EditorTextWithOffset
> = (value): value is EditorTextWithOffset => {
  const { editorText } = value as EditorTextWithOffset;
  return (
    editorText != null &&
    isEditorText(editorText) &&
    typeof (value as EditorTextWithOffset).offset === 'number'
  );
};

interface EditorReactElementFactory<T, P> extends EditorElement {
  readonly tag: T;
  readonly props?: P;
  readonly children: (EditorReactElement | EditorText)[];
}

export type EditorReactElement = $Values<
  {
    [T in keyof ReactDOM]: EditorReactElementFactory<
      T,
      ReturnType<ReactDOM[T]>['props']
    >;
  }
>;

export function materializeEditorPath(
  path: EditorPath,
): (element: EditorElement) => Option<MaterializedEditorPath> {
  return element => {
    const parents: MaterializedEditorPath['parents'] = [];
    let to: MaterializedEditorPath['to'] = element;
    for (let i = 0; i < path.length; i++) {
      const pathIndex = path[i];
      if (isEditorElement(to)) {
        parents.push(to);
        to = to.children[pathIndex];
        if (to == null) return none;
      } else if (isEditorText(to)) {
        const pathContinues = i < path.length - 1;
        if (pathContinues || pathIndex > to.text.length) return none;
        return some({ parents, to: { editorText: to, offset: pathIndex } });
      }
    }
    return some({ parents, to });
  };
}

/**
 * Map `<div>a</div>` to `{ id: id(), tag: 'div', children: [{ id: id(), text: 'a' }] }` etc.
 */
export function jsx(element: JSX.Element): EditorReactElement {
  const {
    type: tag,
    props: { children = [], ...props },
  } = element;
  const editorChildren = Children.toArray(children).map(child => {
    if (typeof child === 'string') {
      const text: EditorText = { id: id(), text: child };
      return text;
    }
    if (child.type === 'br') {
      const text: EditorText = { id: id(), text: '' };
      return text;
    }
    return jsx(child);
  });
  const editorProps = Object.keys(props).length > 0 ? props : undefined;
  return {
    id: id(),
    tag,
    props: editorProps,
    children: editorChildren,
  };
}

export const editorElementChildIsEditorTextNotBR: Refinement<
  EditorElementChild,
  EditorText
> = (child): child is EditorText => {
  return isEditorText(child) && !editorTextIsBR(child);
};

/**
 * Like https://developer.mozilla.org/en-US/docs/Web/API/Node/normalize,
 * except strings can be empty. Empty strings are rendered as BR.
 * If nothing has been normalized, the same element is returned.
 */
export const normalizeEditorElement: Endomorphism<EditorElement> = element => {
  // This flag is good enough for now. We can use fp-ts These later.
  let somethingHasBeenNormalized = false;
  const children = element.children.reduce<(EditorElementChild)[]>(
    (array, child) => {
      if (isEditorElement(child)) {
        const normalizedChild = normalizeEditorElement(child);
        if (normalizedChild !== child) somethingHasBeenNormalized = true;
        return [...array, normalizedChild];
      }
      if (editorTextIsBR(child)) return [...array, child];
      return pipe(
        last(array),
        chain(fromPredicate(editorElementChildIsEditorTextNotBR)),
        fold(
          () => [...array, child],
          previousText => {
            somethingHasBeenNormalized = true;
            return unsafeUpdateAt(
              array.length - 1,
              { ...previousText, text: previousText.text + child.text },
              array,
            );
          },
        ),
      );
    },
    [],
  );
  // Preserve identity, otherwise it would always create new objects.
  // https://github.com/gcanti/fp-ts/issues/976
  if (!somethingHasBeenNormalized) return element;
  return { ...element, children };
};

/**
 * Like https://developer.mozilla.org/en-US/docs/Web/API/Node/normalize,
 * except strings can be empty. Empty string is considered to be BR.
 */
export const editorElementIsNormalized: Predicate<EditorElement> = element => {
  const normalizedElement = normalizeEditorElement(element);
  // We don't need short circuit. We can leverage identity check.
  return element === normalizedElement;
};

// TODO: Can we recursively remove ID from EditorElement type?
export const recursiveRemoveID = (element: EditorElement): any => {
  if (element == null) return element;
  // eslint-disable-next-line no-shadow, @typescript-eslint/no-unused-vars
  const { id, ...objectWithoutID } = element;
  return {
    ...objectWithoutID,
    children: element.children.map(child => {
      if (isEditorText(child)) {
        // eslint-disable-next-line no-shadow, @typescript-eslint/no-unused-vars
        const { id, ...childWithoutID } = child;
        return childWithoutID;
      }
      return recursiveRemoveID(child);
    }),
  };
};

// Functional optics.
// https://github.com/gcanti/monocle-ts

/**
 * Focus on the children of EditorElement.
 */
export const childrenLens = Lens.fromProp<EditorElement>()('children');

/**
 * Focus on the child at index of EditorElementChild[].
 */
export function getChildAtOptional(index: number) {
  return indexArray<EditorElementChild>().index(index);
}

/**
 * Focus on EditorElement of EditorElementChild.
 */
export const elementPrism = Prism.fromPredicate(
  editorElementChildIsEditorElement,
);

/**
 * Focus on EditorText of EditorElementChild.
 */
export const textPrism = Prism.fromPredicate(editorElementChildIsEditorText);

/**
 * Focus on EditorElement by EditorPath.
 */
export function getElementTraversal(path: EditorPath) {
  return path.reduce(
    (acc, pathIndex) => {
      return acc
        .composeLens(childrenLens)
        .composeOptional(getChildAtOptional(pathIndex))
        .composePrism(elementPrism);
    },
    elementPrism.asOptional() as Optional<EditorElement, EditorElement>,
  );
}

/**
 * Focus on EditorText by EditorPath.
 */
export function getTextTraversal(
  parentElementPath: EditorPath,
  index: number,
): Optional<EditorElement, EditorText> {
  return getElementTraversal(parentElementPath)
    .composeLens(childrenLens)
    .composeOptional(getChildAtOptional(index))
    .composePrism(textPrism);
}

/**
 * Ensure text traversal. If EditorPath focuses to text offset, get parent path.
 */
export function ensureTextTraversal(
  path: EditorPath,
  element: EditorElement,
): Optional<EditorElement, EditorText> {
  return pipe(
    path,
    foldRight(
      () => {
        throw new Error('TODO: Refactor.');
      },
      (path, index) => {
        const textTraversal = getTextTraversal(path, index);
        if (textTraversal.asFold().getAll(element).length > 0)
          return textTraversal;
        return ensureTextTraversal(path, element);
      },
    ),
  );
}

export function setTextElement(
  text: string,
  selection: EditorSelection,
): Endomorphism<EditorElement> {
  return element => {
    const onlyTextIsSelected: Predicate<EditorSelection> = selection => {
      const anchorFold = ensureTextTraversal(
        selection.anchor,
        element,
      ).asFold();
      const focusFold = ensureTextTraversal(selection.focus, element).asFold();
      if (!anchorFold.exist(isEditorText)(element)) return false;
      if (!focusFold.exist(isEditorText)(element)) return false;
      return anchorFold.getAll(element)[0] === focusFold.getAll(element)[0];
    };

    if (
      editorSelectionIsCollapsed(selection) ||
      onlyTextIsSelected(selection)
    ) {
      const path = selection.anchor;
      return ensureTextTraversal(path, element).modify(editorText => {
        return { ...editorText, text };
      })(element);
    }
    return element;
  };
}

export function deleteContentElement(
  selection: EditorSelection,
): Endomorphism<EditorElement> {
  return element => {
    const range = editorSelectionAsRange(selection);
    // TODO: Refactor all.
    const anchorMaterializedPath = toNullable(
      materializeEditorPath(range.anchor)(element),
    );
    const focusMaterializedPath = toNullable(
      materializeEditorPath(range.focus)(element),
    );
    if (anchorMaterializedPath == null || focusMaterializedPath == null)
      return element;
    // TODO: Handle other cases, with lenses.
    if (
      // Just deleting text on the same EditorTexts.
      isEditorTextWithOffset(anchorMaterializedPath.to) &&
      isEditorTextWithOffset(focusMaterializedPath.to) &&
      anchorMaterializedPath.to.editorText ===
        focusMaterializedPath.to.editorText
    ) {
      return pipe(
        init(range.anchor),
        fold(
          () => element,
          foldRight(
            () => element,
            (init, index) => {
              // @ts-ignore Refactor later.
              const startOffset = anchorMaterializedPath.to.offset;
              // @ts-ignore Refactor later.
              const endOffset = focusMaterializedPath.to.offset;
              return getTextTraversal(init, index).modify(editorText => {
                const text =
                  editorText.text.slice(0, startOffset) +
                  editorText.text.slice(endOffset);
                return { ...editorText, text };
              })(element);
            },
          ),
        ),
      );
    }
    return element;
  };
}
