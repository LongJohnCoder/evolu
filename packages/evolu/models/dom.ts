/* eslint-env browser */
import { Predicate, Refinement } from 'fp-ts/lib/function';
import { fromNullable, Option, none, some, chain } from 'fp-ts/lib/Option';
import { IO } from 'fp-ts/lib/IO';
import { pipe } from 'fp-ts/lib/pipeable';
import {
  DOMElement,
  DOMNode,
  DOMRange,
  DOMSelectionMaybeNeverExisted,
  DOMText,
  DOMSelection,
} from '../types/dom';
import { DOMNodeOffset, DOMTextOffset } from '../types';
import { unwrapPathIndex, pathIndex } from './path';

export const isDOMElement: Refinement<DOMNode, DOMElement> = (
  node,
): node is DOMElement => node.nodeType === Node.ELEMENT_NODE;

export const isDOMText: Refinement<DOMNode, DOMText> = (
  node,
): node is DOMText => node.nodeType === Node.TEXT_NODE;

export const isDOMSelection: Refinement<
  DOMSelectionMaybeNeverExisted,
  DOMSelection
> = (selection): selection is DOMSelection =>
  selection.anchorNode != null && selection.focusNode != null;

/**
 * Chrome has a bug when it can return obsolete DOMSelection. It happens when a long
 * text is deleted by holding delete key, so it's probably a race condition, because
 * Safari is ok. Therefore, we have to validate DOMNodeOffset to prevent:
 * `Uncaught DOMException: Failed to execute 'setEnd' on 'Range'`
 */
export const isValidDOMNodeOffset: Predicate<DOMNodeOffset> = ([
  node,
  offset,
]) =>
  isDOMElement(node)
    ? unwrapPathIndex(offset) <= node.childNodes.length
    : isDOMText(node)
    ? unwrapPathIndex(offset) <= node.data.length
    : false;

export const getDOMRangeFromInputEvent = (
  event: InputEvent,
): Option<DOMRange> =>
  // @ts-ignore Outdated types.
  fromNullable(event.getTargetRanges()[0]);

export const isCollapsedDOMSelectionOnTextOrBR: Predicate<DOMSelection> = selection =>
  selection.isCollapsed &&
  (isDOMText(selection.focusNode) ||
    selection.focusNode.childNodes[selection.focusOffset].nodeName === 'BR');

export const getTextContentFromRangeStartContainer: (
  range: DOMRange,
) => Option<string> = range => fromNullable(range.startContainer.textContent);

export const preventDefault = (event: InputEvent): IO<void> => () => {
  event.preventDefault();
};

export const DOMSelectionToDOMTextOffset = (
  selection: DOMSelection,
): Option<DOMTextOffset> =>
  pipe(
    pathIndex(selection.anchorOffset),
    chain(offset =>
      selection.isCollapsed && isDOMText(selection.anchorNode)
        ? some([selection.anchorNode, offset])
        : none,
    ),
  );

export const isMoveWithinDOMTextOffset = (
  forward: boolean,
): Predicate<DOMTextOffset> => ([node, offset]) =>
  forward
    ? unwrapPathIndex(offset) < node.data.length
    : unwrapPathIndex(offset) > 0;

export const isDOMTextToBeDeletedByRange: (
  node: DOMText,
) => Predicate<DOMRange> = node => range =>
  range.startOffset === 0 && range.endOffset === node.data.length;

export const isInline: Predicate<CSSStyleDeclaration> = declaration =>
  declaration.display === 'inline';
