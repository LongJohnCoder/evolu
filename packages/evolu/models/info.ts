import { empty, isNonEmpty } from 'fp-ts/lib/Array';
import { absurd, constVoid } from 'fp-ts/lib/function';
import { snoc } from 'fp-ts/lib/NonEmptyArray';
import { geq } from 'fp-ts/lib/Ord';
import { pipe } from 'fp-ts/lib/pipeable';
import { fold } from 'fp-ts/lib/Option';
import { Element, Info, Node, NodeInfo, Path, Selection } from '../types';
import { isElement } from './element';
import { selectionToRange } from './selection';
import { byContains, toPathIndex } from './path';
import { isText } from './text';

type SearchingFor = 'start' | 'end' | 'nothing';

// TODO: Measure then optimize via memoization.
export const createInfo = (selection: Selection, element: Element): Info => {
  let sf: SearchingFor = 'start';
  const range = selectionToRange(selection);
  const nodes: NodeInfo[] = [];

  const walk = (node: Node, path: Path) => {
    if (isNonEmpty(path)) {
      switch (sf) {
        case 'start': {
          // console.log(path, range.start);
          if (isText(node) && geq(byContains)(path, range.start)) {
            // sf = 'end';
            sf = 'nothing';
          }
          break;
        }
        case 'end':
          if (geq(byContains)(path, range.end)) sf = 'nothing';
          break;
        case 'nothing':
          return;
        default:
          absurd(sf);
      }
      const selected = sf === 'end' || sf === 'nothing';

      if (selected) {
        nodes.push({ node, path });
      }
    }

    if (sf === 'nothing' || !isElement(node)) return;
    node.children.forEach((nodeChild, index) => {
      if (sf === 'nothing') return;
      pipe(
        toPathIndex(index),
        fold(constVoid, index => {
          walk(nodeChild, snoc(path, index));
        }),
      );
    });
  };

  walk(element, empty);

  // eslint-disable-next-line no-console
  // console.log(JSON.stringify(nodes));

  return {
    // selection,
    // range,
    nodes,
  };
};
