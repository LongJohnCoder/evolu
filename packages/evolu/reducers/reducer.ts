import { absurd } from 'fp-ts/lib/function';
import { pipe } from 'fp-ts/lib/pipeable';
import { select, setText, deleteContent, setFocus } from '../models/value';
import { Reducer } from '../types';

/**
 * Reducer maps browser actions to Value endomorphisms.
 */
export const reducer: Reducer = (value, action) => {
  switch (action.type) {
    case 'focus':
      return pipe(
        value,
        setFocus(true),
      );

    case 'blur':
      return pipe(
        value,
        setFocus(false),
      );

    case 'selectionChange':
      return pipe(
        value,
        select(action.selection),
      );

    // case 'setText':
    //   return pipe(
    //     value,
    //     // setText(action.operation),
    //   );

    case 'insertText':
      return pipe(
        value,
        setText(action.text),
        select(action.selection),
      );

    case 'insertReplacementText':
      return pipe(
        value,
        setText(action.text),
      );

    case 'deleteText':
      return pipe(
        value,
        select(action.selection),
        setText(action.text),
      );

    case 'deleteContent':
      return pipe(
        value,
        deleteContent(action.selection),
      );

    default:
      return absurd(action);
  }
};
