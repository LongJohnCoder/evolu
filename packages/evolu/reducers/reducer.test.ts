import { editorReducer } from './reducer';
import { createValueWithText } from '../models/value';
import { mapElementToIDless } from '../models/element';
import { Action } from '../types';

const value = createValueWithText();

const testEditorReducer = (action: Action) => {
  const { element, ...nextValue } = editorReducer(value, action);
  return {
    ...nextValue,
    element: mapElementToIDless(element),
  };
};

test('focus', () => {
  expect(testEditorReducer({ type: 'focus' })).toMatchSnapshot();
});

test('blur', () => {
  expect(testEditorReducer({ type: 'blur' })).toMatchSnapshot();
});
