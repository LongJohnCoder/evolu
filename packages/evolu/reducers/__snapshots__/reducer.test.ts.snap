// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blur 1`] = `
Object {
  "element": Object {
    "children": Array [
      "",
    ],
    "props": Object {
      "className": "root",
    },
    "tag": "div",
  },
  "hasFocus": false,
  "info": Object {
    "nodes": Array [],
  },
  "selection": Object {
    "_tag": "None",
  },
}
`;

exports[`focus 1`] = `
Object {
  "element": Object {
    "children": Array [
      "",
    ],
    "props": Object {
      "className": "root",
    },
    "tag": "div",
  },
  "hasFocus": true,
  "info": Object {
    "nodes": Array [],
  },
  "selection": Object {
    "_tag": "None",
  },
}
`;
